Implement single-attribute type map support for SQLData and Enum.

ao-apps opened the pull request:
This is related to issue #641. (https://github.com/pgjdbc/pgjdbc/issues/641)

This includes type inference for `DOMAIN` types that come back from the server as the base type.  Type inference is used in `readObject(String,Class<?>)` and `readObject(int,Class<?>)`, where the type requested is matched to the type map of the connector.

The inference looks for a direct match first, then searches for any inheritance matches.  Thus both providing a means to avoid ambiguity (direct match) and support of the separation of interface from implementation (inheritance match).

When more than one type has been registered for the requested type, an `SQLException` is thrown due to ambiguity.  Please note, however, that inheritance inference will not conflict with direct inference in this base.  Ambiguity is within the categories of direct or inherited.

This also includes support for seamless PostgreSQL `ENUM` mapping to Java `Enum` as `Types.JAVA_OBJECT`.  In the default setting of `typemap`, `Enum` support is only active for types that are in the type map.  When `Enum` support is active, `setObject(…)` (and related) methods will use `Enum.name()` instead of `Enum.toString()`, and `getObject(…)` (and related) methods will use `Enum.valueOf(Class<T>, String)` to retrieve the `Enum` value.

There are some wide-spead, but not complicated, API changes.  Please consider prioritizing this as it will greatly assist us in using `SQLData` and `Enum` for our scalar / non-compound types (`DOMAIN` and `ENUM`).

Thank you.

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing](https://github.com/pgjdbc/pgjdbc/blob/master/CONTRIBUTING.md) document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### New Feature Submissions:

1. [ ] Does your submission pass tests?
2. [x] Does mvn checkstyle:check pass ?

### Changes to Existing Features:

* [x] Does this break existing behaviour? If so please explain.
* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

ao-apps committed:
Implement single-attribute type map support for SQLData in `readObject`.

This is related to issue #641. (https://github.com/pgjdbc/pgjdbc/issues/641)

This is complementary to our simpler pull request #1377 (https://github.com/pgjdbc/pgjdbc/pull/1377) that handles the `writeObject`.

This includes type inference for DOMAIN and ENUM types that come back from the server as the base type.  Type inference is used in `readObject(String,Class<?>)` and `readObject(int,Class<?>)`, where the type requested is matched to the type map of the connector.

The inference looks for a direct match first, then search for any inheritance matches.  Thus both providing a means to avoid ambiguity (direct match) and support of the separation of interface from implementation (inheritance match).

When more than one type has been registered for the requested type, an SQLException is thrown due to ambiguity.  Please note, however, that inheritance inference will not conflict with direct inference in this base.  Ambiguity is within the categories of direct or inherited.

Although there is some API changes, they are generally compatible changes.  Please consider prioritizing this as it will greatly assist us in using SQLData for our scalar / non-compound types (DOMAIN and ENUM).

Thank you.

aoindustries commented:
We are working on build compatibility for the Java 1.6 builds.  Thanks you.

bokken commented on pgjdbc/src/main/java/org/postgresql/jdbc/PgConnection.java:
```diff
@@ -1241,9 +1322,71 @@ public DatabaseMetaData getMetaData() throws SQLException {
+    Class<?> current = clazz;
+    do {
+      classes.add(current);
+      for (Class<?> iface : clazz.getInterfaces()) {
```
what about interface hierarchy? For example, ArrayList implements List which extends Collection. I believe that your set would include ArrayList and List, but not Collection. (Actually it would, but because ArrayList ends up extending AbstractCollection).

aoindustries replied:
Good point.  Thank you.  I'll work on it.

aoindustries replied:
I have reworked this to a recursive implementation to get all interfaces, and extended interfaces, all the way up.  It also adds to the inferenceMap directly, instead of building the intermediate set of classes.

https://github.com/pgjdbc/pgjdbc/pull/1378/commits/3249827059d113bcdd9694a7213cc77809e0cbe3

aoindustries replied:
I have added a test case for the scenario you describe, with a rather contrived interface and class hierarchy: https://github.com/pgjdbc/pgjdbc/pull/1378/commits/72597e4b262cb16284350b89993ccb2fd2ab44f4

bokken commented on pgjdbc/src/main/java/org/postgresql/jdbc/PgConnection.java:
```diff
@@ -1241,9 +1322,71 @@ public DatabaseMetaData getMetaData() throws SQLException {
+    // Make each type set unmodifiable
+    makeValuesUnmodifiable(newDirectMap);
+    makeValuesUnmodifiable(newInheritedMap);
+    synchronized (typemapLock) {
```
what is the purpose of this synchronization? the corresponding reads (from get methods) are not synchronized. 

Is the concern trying to handle concurrent calls to setTypeMap on the same PgConnection?

vlsi replied:
I agree with @bokken : the synchronization should be either in both reads and writes or it should be removed completely.
It makes little sense to synchronize writes only.

ao-apps committed:
Removed use of <> generics for Java 1.6 support.

ao-apps committed:
Changed to fully qualified class name in Javadocs to avoid otherwise unnecessary import.

ao-apps committed:
SQLData.readSQL implementation uses SQLInputImpl, which is available in JDBC >= 4.1.

ao-apps committed:
Now adding all extended interfaces, too.

Also adding directly to inferenceMap instead of building an intermediate set of classes.

ao-apps committed:
@implSpec fails in Java 9

ao-apps committed:
Resolved unreachable statement in JDBC 4.0 build.

ao-apps committed:
Improved accuracy of comment describing our inference-based workaround.

ao-apps committed:
Removed unnecessary cast.

ao-apps committed:
Resolved IDE editor error due to IDE not understanding #if/#endif preprocessing.

Using `if(true)` in `#else` block to always throw exception while still avoiding unreachable code error.

vlsi commented on pgjdbc/src/main/java/org/postgresql/jdbc/PgConnection.java:
```diff
@@ -379,9 +402,19 @@ public CallableStatement prepareCall(String sql) throws SQLException {
   @Override
   public Map<String, Class<?>> getTypeMap() throws SQLException {
     checkClosed();
+    return new HashMap<String, Class<?>>(getTypeMapNoCopy());
```
What's the point of adding a copy here?

aoindustries replied:
Per https://docs.oracle.com/javase/8/docs/api/java/sql/Connection.html#getTypeMap--

"You must invoke setTypeMap after making changes to the Map object returned from getTypeMap as a JDBC driver may create an internal copy of the Map object passed to setTypeMap"

This implies that the returned Map must be modifiable.

And the tutorial shows the use-case of it being always modifiable:
https://docs.oracle.com/javase/tutorial/jdbc/basics/sqlcustommapping.html#using_connection_type_map

davecramer replied:
What about using creating a copy on setTypeMap using unmodifiableMap ?
Note it does say "may" create an internal copy, candidly I presume anyone making a change would call setTypeMap so it's sort of pointless, but if we do it on the setter it gets rid of all the internal copying

vlsi commented on pgjdbc/src/main/java/org/postgresql/core/BaseConnection.java:
```diff
@@ -64,19 +66,86 @@ ResultSet execSQLQuery(String s, int resultSetType, int resultSetConcurrency)
+
+  /**
+   * Gets the unmodifiable map used for type inference of custom types.
+   * This is required because the server returns base types for domains.
```
I'm afraid the description does not clarify what is stored in the map.

Would you please elaborate?

aoindustries replied:
We have reworked this part to be less about "inference map" and more about "inverted map", which better represents what it is.  Please see commit for more details: https://github.com/pgjdbc/pgjdbc/pull/1378/commits/cd9da18e04675dab24765109682d6f8c7cbda011

davecramer commented on pgjdbc/src/main/java/org/postgresql/core/BaseConnection.java:
```diff
@@ -64,19 +66,86 @@ ResultSet execSQLQuery(String s, int resultSetType, int resultSetConcurrency)
+   * {@inheritDoc}
+   *
+   * <p>
+   * Performs a defensive copy, per specification that returned map should be
```
I cannot see where the map needs to be modifiable in the spec ?

aoindustries replied:
Per https://docs.oracle.com/javase/8/docs/api/java/sql/Connection.html#getTypeMap--

"You must invoke setTypeMap after making changes to the Map object returned from getTypeMap as a JDBC driver may create an internal copy of the Map object passed to setTypeMap"

This implies that the returned Map must be modifiable.

And the tutorial shows the use-case of it being always modifiable:
https://docs.oracle.com/javase/tutorial/jdbc/basics/sqlcustommapping.html#using_connection_type_map

ao-apps committed:
Stopping class hierarchy traversals when a class already visited.

This keeps the implementation away from combinatorial explosion on complicated or contrived interface hierarchies.

ao-apps committed:
getObjectImpl did not need to exist - just be in getObject.

ao-apps committed:
Fixed checkstyle.  Old habits die hard.

ao-apps committed:
Cleaning up javadoc warnings; they are considered build errors.

ao-apps committed:
Improving API, naming, and descriptions of the inverted type maps used for type inference.

Change named "inferenceMap" to "invertedMap" to better represent what they are.

Deferred creating of inverted maps until needed.

Removed synchronization on typemap.  Instead synchronizing independently on each inverted maps, giving them a chance to detect a replacement of typemap and recreate their inverted forms.

No longer returning inverted maps, but instead changed the methods to return a `Set<String>` of matching types, or an empty set when there are no types matching the given class.  By not returning a map, the implementation is more free to change as-needed in the future.

aoindustries commented:
> @aoindustries can you add test cases as well ?

We are working on some test cases.

ao-apps committed:
Missed this file on last commit.  Already pushed so not amending last commit.

ao-apps committed:
Ignoring our nb-configuration.xml files, too

ao-apps committed:
Beginning test implementations.

We want to make sure a few simple tests are running getting into more complicate things.

We will also consider working on being able to run the tests against a local PostgreSQL database, instead of relying on the slow iterations through travis-ci.

ao-apps committed:
A bit more testing of the inverted inheritance.

Noted a couple more tests TODO.

ao-apps committed:
Fixed race condition where typemap and its inverted forms could become inconsistent.

ao-apps committed:
LinkedHashMap to maintain order of map provided by caller.

Not part of specification, but is nice to provide this consistency to users.

ao-apps committed:
Added TODO, indicating we still need to implement type mapping here.

ao-apps committed:
Added missing  + 1 on initial hashmap sizing.

Also applied same calculation to the other per-field HashMap.

I know this is off-topic for this pull request.  Maybe this warrants its own tiny pull request, because it is simple but could avoid an occasional rehash where one was not expected (due to (int) cast rounding down).  This is in a part of the code most heavily affecting performance.

ao-apps committed:
Enabled TypeMapTest in Jdbc2TestSuite.

ao-apps committed:
Using isEmpty() instead on null check.

Checking inverted type maps was redefined to return empty set instead of null on no match.

ao-apps committed:
Fixed typo in comment.

ao-apps committed:
Type inference implementation and comment improvements.

Avoiding accessing typemap a second time, since we already have it.

Avoiding what would have been a NullPointerException during race conditions of concurrent type additions.

ao-apps committed:
Renamed "inferredType" to "inferredPgType" to make it clear this is related to "pgType" above.

ao-apps committed:
Avoiding isEmpty() immediately followed by size().

Put common case of size == 1 first.

ao-apps committed:
Per the documentation, a null map will cause the connection's typemap to be used.

https://docs.oracle.com/javase/tutorial/jdbc/basics/sqlcustommapping.html#using_your_own_type_map
"If you do not pass a type map to a method that can accept one, the driver will by default use the type map associated with the connection."

ao-apps committed:
Throwing SQLFeatureNotSupportedException where appropriate for JDBC < 4.1

ao-apps committed:
Including Object.class as a base class for interfaces.

Class.getSuperclass() returns null for interfaces, but interfaces can be cast to Object
For sake of completeness, add Object for interfaces, too

We have tests running locally now, so should clutter up the travis-ci build logs a bit less.

ao-apps committed:
Added missing import required for javadoc.

ao-apps committed:
Added some missing javadocs.

ao-apps committed:
Added some missing javadocs.

ao-apps committed:
Added more tests.

ao-apps committed:
Fixed missing return in JDBC < 4.1 builds.

ao-apps committed:
Tests that interfaces extended by interfaces implemented by base classes are included.

ao-apps committed:
Accepting null on setTypeMap and treating as an empty map.

ao-apps committed:
Separated the tests for setTypeMap and getTypeMap from the inverted map tests.

Created a new test suite for user-defined data types, since there will be quite a few tests.

ao-apps committed:
Space after if ... old habits die hard.

ao-apps committed:
Added some logging to aid in testing and debugging.

ao-apps committed:
Added test of a user-defined data type implemented as a domain over integer.

This test is failing due to implementation via javax.sql.rowset.serial.SQLInputImpl simply performing casts.  We'll look into a more advanced solution, leveraging the existing codebase for the type conversions.

This will also be one step toward removing the dependency on Java 7 for the UDT implementation.

bokken commented on pgjdbc/src/main/java/org/postgresql/jdbc/PgConnection.java:
```diff
@@ -378,10 +384,170 @@ public CallableStatement prepareCall(String sql) throws SQLException {
+  /**
+   * The set of types directly mapped to each class - used for type inference.
+   */
+  private final Map<Class<?>, Set<String>> typemapInvertedDirect = new HashMap<Class<?>, Set<String>>();
```
It looks like we have a mess of variables interspersed through this class :(
@vlsi, @davecramer how do you want to handle this?

aoindustries replied:
We will fix our contributions to be in compliance.

aoindustries replied:
We've fixed the parts we've touched:
https://github.com/pgjdbc/pgjdbc/pull/1378/commits/16293e99e2b60de9208603858ec4d75655563ff5

bokken commented on pgjdbc/src/main/java/org/postgresql/jdbc/PgConnection.java:
```diff
@@ -378,10 +384,170 @@ public CallableStatement prepareCall(String sql) throws SQLException {
+    //#endif
     checkClosed();
+    // LinkedHashMap to maintain order of map provided by caller.
+    return new LinkedHashMap<String, Class<?>>(getTypeMapNoCopy());
```
if the expectation is that the driver is maintaining a defensive copy, why do extra work to preserve order?

aoindustries replied:
Just to be nice to the API user.  So if they are debugging, what goes in matches what comes out later.  Makes things easier to read from debug log files, for example.

bokken replied:
so if a consumer passes in a `TreeMap` with a `String.CASE_INSENSITVE_ORDER` we should be concerned about preserving iteration order even though we do not care about changing the behavior of the map to operate on `hashCode` and `equals` rather than a provided comparator?

Absent using a `TreeMap`, it is unclear to me why consumers would care about iteration order.

aoindustries replied:
Most users wouldn't care.  But why not maintain the order?  No harm done.

Keep in mind, that the tutorials show to call getTypeMap(), make any changes, then call setTypeMap() again.  Now people can call setTypeMap() with any map they want, not necessarily the one they got from us.  But if they follow this common convention, they map will maintain the order they added the types.  This can be useful and comes at extremely little cost.

Ultimately, I'm doing this for ourselves.  We register quite a few custom types in our applications.  When I'm digging through log files debugging a problem, having consistent ordering of things helps, instead of stuff jumbled all around during a rehash.  This can make comparing different log lines via `diff --color` readily show the differences in the types mapped.  Sure, I could sort the maps when logging, but why not get this for free from the API?

ao-apps committed:
Implemented SQLInput from ResultSet.

This removes the dependency on Java 7.
This fixes the broken test case for the Port domain-over-integer test.

ao-apps committed:
Inlined return with ternary operator.

aoindustries commented on pgjdbc/src/main/java/org/postgresql/jdbc/PgConnection.java:
```diff
@@ -378,10 +384,170 @@ public CallableStatement prepareCall(String sql) throws SQLException {
+      }
+      types = typemapInvertedDirect.get(clazz);
+    }
+    if (types == null) {
```
Fixed.  We had done this to avoid the funky `Collections.<String>emptySet()` syntax required for Java 6.

https://github.com/pgjdbc/pgjdbc/pull/1378/commits/ae042fac86ad45099a8472ecaa4f1495cee8a042

ao-apps committed:
SQLFeatureNotSupportedException is no longer thrown for JDBC < 4.1.

ao-apps committed:
Reverted comment to minimize the diff from master.

ao-apps committed:
Moved fields to top of class.

bokken commented on pgjdbc/src/main/java/org/postgresql/jdbc/PgConnection.java:
```diff
@@ -378,10 +384,170 @@ public CallableStatement prepareCall(String sql) throws SQLException {
+          addAllInverted(invertedMap, iface, type);
+        }
+        // Class.getSuperclass() returns null for interfaces, but interfaces can be cast to Object
+        // For sake of completeness, add Object for interfaces, too
```
should Object be ignored entirely?

aoindustries replied:
The inverted maps now skip Object.  However, a custom type could still be directly mapped to Object:

https://github.com/pgjdbc/pgjdbc/pull/1378/commits/d1a0e9e0c055b0f9e4733ca0c3effc249825181b

bokken commented on pgjdbc/src/main/java/org/postgresql/jdbc/PgResultSet.java:
```diff
@@ -3432,11 +3395,89 @@ public void updateArray(String columnName, Array x) throws SQLException {
+      if (!typemap.isEmpty()) {
+        String pgType = getPGType(columnIndex);
+        Class<?> customType = typemap.get(pgType);
+        if (customType != null) {
```
should there be a comparison between the provided `type` and `customType`? the current behavior will result in a ClassCastException, where it seems a SQLException with CANNOT_COERCE state would be better.

aoindustries replied:
Fixed.  Thanks.
https://github.com/pgjdbc/pgjdbc/pull/1378/commits/5ff58d39e746a0802ca03a9a9a6ac46a09de9997

ao-apps committed:
Added SQLData support to PreparedStatement.

ao-apps committed:
Now using PgResultSet and PgPreparedStatement directly.

The base ResultSet and PreparedStatement are missing some methods in JDBC < 4.1.  By using these Pg variants, we do not need to #if/#endif around these limitations.

Also, by moving these classes into this package, we were able to make fewer of the methods public, thus cluttering the public API less.

ao-apps committed:
Throwing SQLException instead of ClassCastException when mapped type incompatible with requested type.

ao-apps committed:
No longer building inverted maps for Object.

Type inference to "Object" is of limited utility, so probably not worth maintaining the mapping from all registered types back to Object.class.

ao-apps committed:
Checking Logger.isLoggable before creating Object[] parameters.

ao-apps committed:
Added more tests of custom types, null states, overriding base types, ...

ao-apps committed:
Fixed null return on getObject for both PGobject and SQLData.

ao-apps committed:
SQLData.getSQLTypeName() does not necessarily match the pgType mapped to the custom type.

This did not match when we added a custom type mapping from "int4" to PortImpl or from "text" to Email.

ao-apps committed:
SQLType was added in JDBC 4.2.

ao-apps committed:
These tests are for JDBC >= 4.1

ao-apps committed:
These methods exist here, but are only used from the org.postgresql.jdbc package.

Should they only exist on PgConnection?
Are these intended to be part of a stable API?

ao-apps committed:
Working on SQLInput and SQLOutput implementations.

ao-apps committed:
-1 used instead of Types.LONGVARCHAR - is this expected?

ao-apps committed:
Mixing up variants of methods called for more test coverage.

ao-apps committed:
Testing user-defined data types may only have a single attribute, and must have that one attribute.

ao-apps committed:
Checking wasNull not allowed before first read.

ao-apps committed:
Testing the limits of one SQLData type reading/writing nested SQLData attribute.

Currently, writing works correctly, but reading fails for a couple of different reasons.  First, readObject without a type parameter has unbounded recursion between itself and getObject.  Second, readObject with a type parameter suffers the ambiguity of not knowing what to get from the type map.

This is beyond the scope of our current goals of getting simple DOMAIN and ENUM working well for user-defined data types.

ao-apps committed:
These tests are for JDBC >= 4.1

ao-apps committed:
Testing type map parameter overrides connection's type map.

ao-apps committed:
Trying to fix openjdk7 builds.

ao-apps committed:
Implemented getSQLTypeName() to properly return the typeName given to readSQL(...)

ao-apps committed:
Another "TODO" of where type map might apply.

ao-apps committed:
Trying to fix openjdk7 builds.

ao-apps committed:
These tests are for JDBC >= 4.2

ao-apps committed:
Implemented custom types for getObject in CallableStatement.

Because the specifics of how getObject will be called is not known at the time of `executeWithFlags(int)`, I had to hold onto the `ResultSet` and delegate to it for `getObject` methods.

With this result set available, I was able to delegate the other get* methods to it, which is more direct than the old method of getting everything as objects, storing in an array, and casting each object on the way out.

Also with the result set available, I implemented the remaining get* methods through simple delegation to ResultSet.

In order to access the type inference of `PgResultSet.getObjectImpl` while delegating to `ResultSet`, I had to change a few places the be `PgResultSet` instead of just `ResultSet`.

There are some "TODO" notes left where I'm not immediately sure what to do.

All of the existing tests pass locally.  We will work on tests for the funtionality we have updated or added.

ao-apps committed:
Fixed to use customType.

ao-apps committed:
getObjectImpl will not be necessary if we can continue to delegate to PgResultSet.

ao-apps committed:
Clearer comment wording.

ao-apps committed:
Storing callResultColumnIndex 1-based instead of 0-based.

ao-apps committed:
StandardCharsets is Java 7+

ao-apps committed:
UnsupportedCharsetException might only happen in Java 6.

ao-apps committed:
TODO: We may need to check custom map sooner to detect when requested type is incompatible with mapped type.

ao-apps committed:
Noted documentation source in deciding to use the connection's map when a null map is provided to either ResultSet or CallableStatement.

The Javadocs do not specify this behavior directly, but we are going by https://docs.oracle.com/javase/tutorial/jdbc/basics/sqlcustommapping.html:

    "If you do not pass a type map to a method that can accept one, the driver will by default use the type map associated with the connection."

This is in contrast to the Javadoc definition for Array.getArray(Map) https://docs.oracle.com/javase/8/docs/api/java/sql/Array.html#getArray-java.util.Map-

    "This version of the method getArray uses either the given type map or the standard mapping; it never uses the type map associated with the connection."

ao-apps committed:
Implemented user-defined data types for Arrays.

We are trying to keep the streamlined array creation in PgArray as it has been, while extending it to support custom types.  To this end, we have created an abstraction `ValueAccess`, with implementations that get values from various places, such as `PgResultSet` or values directly available during `PgArray.getArray(...)`.

We then pulled the code from `PgResultSet.getObjectImpl` into a helper class `ValueAccessHelper`, that creates objects from any `ValueAccess` (either from `PgResultSet` or direct values during array creation).

We also moved the management of the type map, along with its inverted forms, into a new class `UdtMap`.  Previously, type inference was only performed based on the typemap of the connection.  With this change, the current typemap is used for inference, whether obtained from the connection, the current `getObject(Map)` invocation, the current `Array.getArray(Map)` invocation, or access to the `ResultSet` obtained from an invocation to `Array.getResultSet(Map)`.

That last point is worth clarifying: When a `ResultSet` comes from an `Array` via `Array.getResultSet(Map)`, calls to `ResultSet.getObject(...)` without a map will use the type map from `Array.getResultSet(Map)`, not the connection's type map.  Same for `Array.getResultSet()` without map - uses the type map associated with the array, not the connection, when not specific map provided to the `ResultSet.getObject(...)` methods.

We have scattered quite a few "TODO" notes around.  We will review these later and, if relevant, create separate, smaller pull requests addressing smaller issues found in the code (such as `Array.getArray(...)` and `Array.getResultSet(...)` with a count of `0` returning all elements.  We're not sure if this is part of the spec.  Plus quite a few places where int overflows can happen without being caught by the current implementation, as well as unsafe casts to (int) on large indexes in `Array.getArray(...)` and `Array.getResultSet(...)` calls.

Another observation is that the `PgResultSet` uses `PSQLState.DATA_TYPE_MISMATCH` when cannot convert types.  Should we use this for custom type conversion problems, too?

We still have a lot of work to do, but want to commit this now so we can start work on any issues in the different builds for different Java and JDBC versions.  All tests pass locally.

Final point for this commit: We could implement this with fewer changes to the code by dispatching calls to `Array.getArray(...)` for custom types through the `Array.getResultSet(...)` implementation.  This would have all object conversion end up going through the existing implementation in `PgResultSet`.  We will modify our implementation of `PgArray` to support both of these tactics (more direct via `ValueAccess` and less direct via `PgResultSet`) and benchmark the results.  This will help guide the decisions on whether the greater codebase changes implied by `ValueAccess` are acceptable for this pull request.  We do not want to change things so much that this pull request becomes untenable.

ao-apps committed:
Getting ArrayAssistant once outside loop.

ao-apps committed:
UdtMap.EMPTY on empty map, too.

ao-apps committed:
Defer creating new UdtMap until needed.

ao-apps committed:
This comment refers to "implementation" change, not "behavior" change.

ao-apps committed:
@Override here, too.

ao-apps committed:
Reviewed TextValueAccess for consistency with PgResultSet.

ao-apps committed:
Sets within inverted maps are no longer unmodifiable.

Since inverted maps might now be generated per-request when custom types are provided in getObject(Map), minimizing the overhead of the inverted map creation by adding the first element as Collections.singleton, then upgrading to HashSet once a second type is added for a given class.

In the common case of a direct match (no inheritance matching needed), and a single type registered for a given class, this will result in the direct inverted map containing only singleton sets and the inheritance map never being created.

ao-apps committed:
Noted a few more things to test.

ao-apps committed:
Reviewed uses of castToBoolean that can be more direct fromString.

ao-apps committed:
Noted a few more things to test.

ao-apps committed:
UdtMap is more properly part of the SQLInput, not the ValueAccess.

ao-apps committed:
Implemented getObject for ValueAccess, sharing underlying implementation with PgResultSet.

ao-apps committed:
Fixed PSQLState to match previous value.

ao-apps committed:
New UUIDConverter consistent with other converters.

ao-apps committed:
TODO: Locale-sensitive toUpperCase() where it should not be.

ao-apps committed:
This looks like a very good candidate for an Enum.

ao-apps committed:
Got first tests of UDT arrays running.  Fixed one bug.

ao-apps committed:
This is OK.  The additional dimensions are created with length 0.

ao-apps committed:
Added tests for custom type arrays in two dimensions.

ao-apps committed:
Working on a test of large arrays.

ao-apps committed:
Adding timing of test.

ao-apps committed:
Adding ResultSet tests.

ao-apps committed:
Added iteration timing.

ao-apps committed:
Added query timing, so can put array times into perspective.

ao-apps committed:
Added query timing, so can put array times into perspective.

ao-apps committed:
Added first and second access timing.

ao-apps committed:
Added timing of ResultSet.getArray(int)

ao-apps committed:
Noted potential parser slight speed-up.

ao-apps committed:
Fixed checkstyle.

ao-apps committed:
Fix for openjdk7's brain-dead generic method type inference.

ao-apps committed:
Testing if implNote works on all builds.

ao-apps committed:
implNote is a no-go.

ao-apps committed:
Defining wasNull() on a per type basis.

ao-apps committed:
Removed unused SQLException.

ao-apps committed:
Consider renaming to BaseValue/PgValue/Pg...

ao-apps committed:
throws SQLException for consistency

ao-apps committed:
Added Calendar variants of getDate, getTime, and getTimestamp.

ao-apps committed:
Noted another redundant getDefaultCalendar().

ao-apps committed:
Remove getDefaultCalendar() from ValueAccess.  Changed to getCalendar() that returns the actual value needed for getObject(Class).

ao-apps committed:
Implemented getLocalDateTime() on Timestamp values only, matching PgResultSet.

ao-apps committed:
Added missing SQLFeatureNotSupportedException, which helps document this class.

ao-apps committed:
Finished Date, Time, and Timestamp for single-attribute SQLInput.

ao-apps committed:
Added tests of custom types on a domain over timestamptz.

ao-apps committed:
Added pgType to the exception, since it is more immediately helpful than a numeric SQL type.

ao-apps committed:
Added support for binary array representations of date, time, timetz, timestamp, and timestamptz.

However, I do not know if the server has the capability to send these in binary form.  I left a "TODO" note.  If we do not find a correct way to turn-on binary mode for these types, we'll revert this change.

ao-apps committed:
domain over timestamptz, not timestamp

ao-apps committed:
Now allowing custom types to override base types on calls to getObject(Class), too.

We made a custom type that extends BigDecimal.  We then mapping "numeric" to this custom type.

We confirm that that custom type is returned on a call to getObject(BigDecimal.class), which works because the custom type is assignable to BigDecimal.

ao-apps committed:
SQLData now takes precedence on setObject, too.

ao-apps committed:
Testing SQLData interface is only used for SQLType OTHER.

ao-apps committed:
SQLData interface taking priority regardless of SQL type.

Upon closer reading of the Javadocs, it seems SQLData interface should apply to all SQL types, not just OTHER.

ao-apps committed:
Explained why cast needed and not needed for Types.OTHER with and without SQLData.

ao-apps committed:
Testing type inference happens after standard mappings.

So the order is:
1) NULL
2) Custom type mappings
3) Standard type mappings
4) Inferred user-defined type mapping (only required because DOMAIN sends base OID)

ao-apps committed:
Added test for enums.  Tests are failing because enums not yet implemented.

ao-apps committed:
QUERY_MODE=simple does not work without the casts, it returns String

ao-apps committed:
Fixed import.

vlsi commented on pgjdbc/src/test/java/org/postgresql/test/udt/CustomTypePriorityTest.java:
```diff
@@ -0,0 +1,177 @@
+import java.util.Collections;
+import java.util.Map;
+
+public class CustomTypePriorityTest {
```
Just a random thought
1) you might want to use `extends BaseTest4` (to save `setUp` `tearDown`) repetitions.
2) Please add newly added test files to an existing suite. Otherwise the test files are not run in the TravisCI

aoindustries replied:
Thanks.  We had used `BaseTest4` before, we we're sure we should.

We'll include all these tests at the end of `Jdbc2TestSuite`.

aoindustries replied:
Thanks done.  I can confirm at least some of our tests were running, by the way:

https://travis-ci.org/pgjdbc/pgjdbc/builds/472061741?utm_source=github_status&utm_medium=notification
https://travis-ci.org/pgjdbc/pgjdbc/jobs/472061755

Are you sure we cannot have a different test suite?  It sure is convenient to just fire of this suite while developing.

davecramer commented:
@aoindustries 
I'm a little concerned that you are not receiving feedback here. We are all sort of waiting to see when this is done. Somehow it would be good if we could pause and possibly review?

ao-apps committed:
Made tests pass with current implementation,

"TODO" where tests will change when enum supported.

aoindustries commented:
> @aoindustries
> I'm a little concerned that you are not receiving feedback here. We are all sort of waiting to see when this is done. Somehow it would be good if we could pause and possibly review?

Short of more testing and reviewing a bunch of our TODOs, many of which are observations not directly related to the stated objective of single-attribute user-defined data type support, we could now be considered feature-complete per stated goals.

However, goals stated `ENUM` support, as well, to which we're exploring adding Java Enum support (configurable "all"/"none"/"typemap").  I would like "typemap" to be the default setting.  In this mode, the driver would support Java Enum in the custom type map, with `Enum.name()` (instead of `Enum.toString()`) being called during `setObject`, `Enum.valueOf(...)` being called for `getObject`.  This will also use the typemap to avoid having to cast the parameters `?::enumtype` to just `?`, too.

Since the `ENUM` type name is passed back from the server, unlike `DOMAIN`, this will facilitate much more seemless mappings of `Enum` objects to `ENUM` types.

The test shows what we're picturing: https://github.com/pgjdbc/pgjdbc/blob/f01a765604b0963994a560b9e72e6e8421f37b59/pgjdbc/src/test/java/org/postgresql/test/udt/EnumTest.java

ao-apps committed:
Commented test a bit more.

ao-apps committed:
Noted this intentionally does not extend SQLData.

ao-apps committed:
Extending BaseTest4.

ao-apps committed:
Consolidated tests into Jdbc2TestSuite.

ao-apps committed:
Removed old "con" causing test failure.

ao-apps committed:
More assertSame for Enum values.

ao-apps committed:
Removed unused import.

ao-apps committed:
Added tests of enum with type inference.

ao-apps committed:
"inferrence" is "inference"

ao-apps committed:
Added comments on why :: casts are and are not used in the tests.

aoindustries commented:
Merry Chistmas everybody!  I'm off for a big dinner with friends.

ao-apps committed:
Removed quotes from `noquotename`.

ao-apps committed:
Added more bounds checks on `count`.

Also removed "TODO" notes about 32-bit indexes possibly overflowing.  The existing implementation will throw `IndexOutOfBoundsException` in this case, except perhaps in some specially crafted scenario where something could overflow all the way through the negative range.

ao-apps committed:
These tests are JDBC >= 4.1 only.

ao-apps committed:
Removed space from "// #endif"

ao-apps committed:
Renamed methods for clarity.

ao-apps committed:
Still having issues with QUERY_MODE=simple.

ao-apps committed:
Renamed methods for clarity.

ao-apps committed:
getPGType() is different than psql when schema requires quoting and name does not.

ao-apps committed:
Noted possible fix regarding disambiguation of case-sensitive names while still keeping case-insensitive matches.

ao-apps committed:
Noted a couple more places that might be affected by user-defined data types.

ao-apps committed:
Working on handling QUERY_MODE=simple.

ao-apps committed:
Class must be public for JUnit member access in Java 9.

java.lang.IllegalAccessException: Class org.junit.runners.model.FrameworkMethod$1 can not access a member of class org.postgresql.test.udt.BaseConnectionTest with modifiers "public"
	at sun.reflect.Reflection.ensureMemberAccess(Reflection.java:105)
	at java.lang.reflect.Method.invoke(Method.java:613)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	...

ao-apps committed:
Fixed column index for getting column type name.

ao-apps committed:
Testing mapping of "unknown" for QUERY_MODE=simple.

ao-apps committed:
Further commented "TODO" regarding if serial/bigserial should be used in user-defined data type mapping versus int4/int8.

ao-apps committed:
Documented our intended support for Enum in PGProperty.

ao-apps committed:
Merge branch 'issue-641-SQLData-support' of github.com:aoindustries/pgjdbc into issue-641-SQLData-support

ao-apps committed:
Fixed checkstyle.

ao-apps committed:
Line-wrapped at 120 characters to avoid horizontal scrolling in the Github interface.

ao-apps committed:
Stronger definition: "does not conflict" versus "should not conflict".

ao-apps committed:
Using arguments to message to assist translators.

ao-apps committed:
Noted: Consider unbounded ExecutorService here instead of spawning a thread for each connect.

ao-apps committed:
Added getter/setter for "enumMode".

ao-apps committed:
Implemented Enum support for getObject / getArray.

ao-apps committed:
Corrected comment to match implementation.

ao-apps committed:
Implemented Enum support for setObject, but not yet avoiding ::enumtype casts.

ao-apps committed:
Reduced iterations and size of test since not actively benchmarking implementation performance.

ao-apps committed:
"Oid.UNSPECIFIED" is more meaningful than "0"

ao-apps committed:
More detailed logging in FINEST.

ao-apps committed:
Documented desired changes for matching multiple values instead of a single getPGType() result.

ao-apps committed:
Noted could be PgConnection.

ao-apps committed:
Added `Enum` support to `setObject(...)`.

This is a pretty big commit.  Since we can't assume a simple one-to-one relationship between oid and pgType, in order to be able to both call SQLOuput from and create the correct user-defined data types on function return in `CallableStatement`, we are passing the pgType and scale through to the `ParameterList`.

The pgType associated with the parameters is used the return value of `PgParameterMetaData.getParameterTypeName(int)`.

The pgType associated with the parameters is also used, in conjunction with the type map, to return the correct Java class from `PgParameterMetaData.getParameterClassName(int)`.

We are also using this information to provide additional casts within `SimpleParameterList`, which is needed to cast as domain type instead of base type when a user-defined data type is mapped.

We have noted that the assumption of a one-to-one relationship from `oid` back to `pgType` is currently an issue.  We will create a version of `TypeInfo.getPGType(int)` that returns the set of all known representations of the type.  This will give us the necessary information to match types correctly, independent of whether the schema quoted, the name is quoted, and if schema present or on querypath.

The existing code that maps an `oid` back to a single `pgType` will favor a mapped type.  This way, whichever form somebody maps a type (schema, no schema, quoted, not quoted) will be returned as the type name in places like `ParameterMetaData` and `ResultSetMetaData`.

We have left "TODO" notes where scaleOrLength should validate the length of the `InputStream` or `Reader` parameters, per the specification.  Now that the scale is passed through to the parameters, the necessary information is available.  We will not implement this verification on this release.

We noted quite a few more "TODO" of things unfinished or we believe are in error (such as a `CompositeParameterList.setLiteralParameter(...)` calling `subparams.setStringParameter(...)`.  We will dobule-check our understanding of these before asking for review.

We don't want to cause grief within the code, so we do not apply globally, but we have been marking "@Override" where missing on any method we interact with.

`PgPreparedStatement` got a while bunch of `*Impl` methods that take an additional `pgType` parameter (null for the default behavior).  These are necessary so the original `pgType` can remain in effect while `SQLOutput` is writing the object through `PgPreparedStatementSQLOutput`.  We marked the existing methods as `final`, and dispatched to the new `protected` `*Impl` versions of the methods.  Thus anybody who has overridden any of the methods will need to override the `*Impl` version instead, maintaining the current `pgType` as appropriate.  We could have used a `ThreadLocal` to avoid this extra parameter, but we're trying to avoid any "spooky action at a distance" within the code.  By adding the additional parameter it is easy to see what is happening, where from, where to, and why.

We still have a couple remaining issues with `EnumTest` related to inconsistent `pgType` coming back from `TypeInfo`.  This should be resolvable with the support of multiple `pgType` associated with a single `oid`.  The tests are altered to pass, with "TODO" markers where not yet correct.  All tests pass locally.

ao-apps committed:
Consider Enum support being Types.JAVA_OBJECT instead of Types.OTHER

ao-apps committed:
ResultSet.updateObject(…) methods need to support typemap, too.

ao-apps committed:
Noted requirements for `getUDTs` implementation.

ao-apps committed:
Using type from PGobject when the pgType is not yet set.

ao-apps committed:
Using type from "hstore" when the pgType is not yet set.

ao-apps committed:
Consider Enum support being Types.JAVA_OBJECT instead of Types.OTHER

ao-apps committed:
PGobject are handled as extensions to base types and do not participate in the user-defined data types.

However, a UDT could be mapped onto a PGobject, and can read/write it in SQLInput and SQLOutput.

ao-apps committed:
PGobject will stay here.

ao-apps committed:
ENUM requires PostgreSQL 8.3+

ao-apps committed:
Create test schema before checking PostgreSQL version.

The "drop table if exists" in After fails when the schema does not exist.

ao-apps committed:
Removed accidental ";;" and varied styles of clearing typemap.

vlsi commented:
Re typecache, there's https://github.com/pgjdbc/pgjdbc/pull/914
Have you seen it? It does not solve one-one issue though.

aoindustries commented:
> Re typecache, there's #914
> Have you seen it? It does not solve one-one issue though.

Thanks.  That is interesting but I think that goes beyond what we'll need.  The TypeInfoCache already adds multiple `pgType` for a single `oid`.  I think all we'll need to do is create the inverted form `Map<Integer, Set<String>>` of this many-to-one mapping.  Then we'll add a couple more variants regarding quoting.  The current code will quote the name only when the schema needs quoted, for example, which is inconsistent with `psql`.
https://github.com/aoindustries/pgjdbc/blob/1e53974c59a34844fd8a76e43c71515f668a21ef/pgjdbc/src/main/java/org/postgresql/jdbc/TypeInfoCache.java#L416-L433

aoindustries commented:
Hah, I just had to fix a production bug caused by our switching some columns to `ENUM` type, which this current development will solve.  We're really looking forward to the `Enum` support helping us avoid casting.  I had to add all of these `::enumtype` casts to our update to fix our payment forms:

Real-world code.  With Enum support, none of the highlighted casts will be necessary:

`"update\n"
+ "  payment.\"Payment\"\n"
+ "set\n"
+ "  authorization_communication_result=?**::payment.\"TransactionResult.CommunicationResult\"**,\n"
+ "  authorization_provider_error_code=?,\n"
+ "  authorization_error_code=?**::payment.\"TransactionResult.ErrorCode\"**,\n"
+ "  authorization_provider_error_message=?,\n"
+ "  authorization_provider_unique_id=?,\n"
+ "  authorization_provider_approval_result=?,\n"
+ "  authorization_approval_result=?**::payment.\"AuthorizationResult.ApprovalResult\"**,\n"
+ "  authorization_provider_decline_reason=?,\n"
+ "  authorization_decline_reason=?**::payment.\"AuthorizationResult.DeclineReason\"**,\n"
+ "  authorization_provider_review_reason=?,\n"
+ "  authorization_review_reason=?::**payment.\"AuthorizationResult.ReviewReason\"**,\n"
+ "  authorization_provider_cvv_result=?,\n"
+ "  authorization_cvv_result=?**::payment.\"AuthorizationResult.CvvResult\"**,\n"
+ "  authorization_provider_avs_result=?,\n"
+ "  authorization_avs_result=?**::payment.\"AuthorizationResult.AvsResult\"**,\n"
+ "  authorization_approval_code=?,\n"
+ "  capture_time=?::timestamp,\n"
+ "  capture_username=?,\n"
+ "  capture_principal_name=?,\n"
+ "  capture_communication_result=?**::payment.\"TransactionResult.CommunicationResult\"**,\n"
+ "  capture_provider_error_code=?,\n"
+ "  capture_error_code=?**::payment.\"TransactionResult.ErrorCode\"**,\n"
+ "  capture_provider_error_message=?,\n"
+ "  capture_provider_unique_id=?,\n"
+ "  status=?**::payment.\"Payment.Status\"**\n"
+ "where\n"
+ "  id=?\n"
+ "  and status='PROCESSING'"`

ao-apps committed:
Created 1.x branches for SemanticCMS 1 and related projects

ao-apps committed:
Now supporting Java 12.

1) Added new AO Badges.

2) Updated parent POM

   a) Changed default Java version from 1.6 to 1.7

   b) Renamed Maven build property names to not conflict with newer versions of Java:
      *.jdk15 -> *.jdk5
      *.jdk16 -> *.jdk6
      *.jdk17 -> *.jdk7
      *.jdk18 -> *.jdk8

ao-apps committed:
Reverted commit e11400b72fc38e7de0571fe10bb819bca0a5815d

ao-apps committed:
Reverted commit a912efd5c0d8bc841425b50d203843e2ee091ac7

crowmagnumb commented:
What's the status of this one? Seems pretty quiet these days. I've just started implementing UDT's in my db and was trying to use the ResultSet.getObject(String columnLabel, Class<T> type) method and of course getting an unsupported exception. So now I'm using PGJDBC-NG driver to make a separate connection whenever I need this functionality. But this driver has a couple other issues for me but rather than switch over completely anyway I'ld love to know what the status of this one is. I've been using this driver almost since its inception!  ;)

davecramer commented:
@crowmagnumb looks pretty stale.

Care to take a run at cleaning it up?

crowmagnumb commented:
@davecramer Ha. Well I'll see what kind of time I can muster towards that.  :)

taojoe commented:
so what's the status?

taojoe commented:
the server supports composite type, but the client doesn't.  I googled for this, and I found it's not new, it's back to 2006. 

https://www.postgresql.org/message-id/Pine.BSO.4.61.0601080311300.11669%40leary.csoft.net

ao-apps committed:
Ignoring .m2 directory used by per-workspace Jenkins builds

ao-apps committed:
Also ignore *@tmp directories used by Jenkins builds of sub-projects

vlsi commented:
Thank you for all the effort here. I'm closing this in favour of #3062, which covers SQLData and enum mapping as part of a broader type-cache and composite-type rework.
